### DISCUSSION TITLE
Having two keys?

### DISCUSSION THREAD

**Discussion by berlef80** [Chia-Network/chia-blockchain#1116] [Q&A]
Is there any benefit to having more than one private key - aka if you are plotting to two machines, for maybe security purposes would it be wise to have a second key? can you open two keys at once in the Chia GUI if one wanted to combine farming on one computer in the future? Thanks!

**Comment by mariano54** [comment]
There are three concerns here:
1. Your farming keys should not be in your plotting machines. Instead, plot by using the -f and -p flags (public keys), and then transfer the plots to the machine you will use to farm. 
2. Your farming keys should not be in your harvester machine, if you have a multi-machine setup. https://github.com/Chia-Network/chia-blockchain/wiki/Farming-on-many-machines
3. Your reward address for chia rewards should be a separate key as well, kept in an offline machine. You can generate an address on a different computer, and put this address in the config.yaml (farmer.xch_target_address and pool.xch_target_address), so if your farming machine gets hacked, you don't lose past rewards.

More technical details here: https://github.com/Chia-Network/chia-blockchain/wiki/Chia-Keys-Architecture

**Reply by berlef80** [reply]
Interesting - Thank you.  Never would have thought of it that way.  I will dig into this more.,  Right now I am using the windows GUI but it looks like i need to work out how to use commands, to plot with -f and -p flags or can I set that up in the GUI?  I looked and couldn't find a place where it explains how to get to the back end where I would put these flags, for someone just starting out (I know basic coding, but not sure where to start with chia),

So just to clarify one question on the above - so the plots are not related to the keys they were plotted in?  So I can use a plot from with one key (I know you say no keys) and farm from another?
Thanks!

**Reply by mariano54** [reply]
When making a plot you can either give it your private keys (this is the default behaviour), or your public keys. Public keys are derived from the private keys but it is public information which you can share with anyone. 

Plotting with public keys can't be done in the GUI right now, only the command line. In theory you can give someone your public keys, they can plot, and then they can send you the file. Farming it requires the private key.

**Reply by berlef80** [reply]
Coming back to this...I have a few more questions.  (To note:  I am using the PC GUI, so know that is less versatile)
I have two plotting machines using the same key to plot. I save final plots onto an external hard drive.  For the testnet, the farming has been done on the same machines using the same key.
I have a laptop that I could plug the extremal drives into and farm from there using a different key, but that requires moving the new plots over to that computer each time they are finished plotting.  Is this your recommendation vs just plotting off the plotter machines and the same key as plotting?  I assume from the above yes?
For the Chia that are won on the main net, you are saying to transfer them from the farming key to a different third key.  Could this be on the same laptop farming machine or should it be on a completely separate machine?
Thanks in advance for your help with this!!

**Reply by Handri-Kosada** [reply]
> 3. Your reward address for chia rewards should be a separate key as well, kept in an offline machine. You can generate an address on a different computer, and put this address in the config.yaml (farmer.xch_target_address and pool.xch_target_address), so if your farming machine gets hacked, you don't lose past rewards.

Hi,

so I just generate new wallet, and grab the wallet address and put into those 2 target address in config.yaml?
It wont clash with my current farmer / harvester keys?

Thanks for the info.

**Reply by nobodyshome** [reply]
> 
> Hi,
> 
> so I just generate new wallet, and grab the wallet address and put into those 2 target address in config.yaml?
> It wont clash with my current farmer / harvester keys?
> 
> Thanks for the info.

Believe this should answer your question: https://github.com/Chia-Network/chia-blockchain/wiki/Chia-Keys-Management

**Reply by VanQz** [reply]
Need some help understanding the whole chia network structure.

Correct me if im wrong:
1. Main Node (Full Node) - used to sync the blockchain nothing more; Connects to other nodes
      a. Wallet is here
2. Farmer - Connect to harvester where the plots are stored. Used to do "mining" (farming)
3. Harvester - Connects to Farmer only (local net) This is storage array for the plots
4. Plotter - Creates the plots

So to better understand here, All these are installed in a single machine when its a single machine setup However from what i understand from good security practice is to have wallet stored elsewhere so ill generate a wallet it provides 4 sets of keys and 24 words mnemonic.

**4 sets of keys include:**
Master Private/public key
Farmer Private/public key
Pool Private/public key
Wallet Private/public key

So I have 2 machines, 1 to plot and 1 to farm.

The plot machines only plot then stores on the main machine which assumes the full node role (wallet, farm, plot, harvester)
I have plots already and i have never used -f -p when creating plots. What would be the best practice here?

**Reply by PAVoutsinas** [reply]
Is there any sensitive information / ID numbers in the plot log? I shared a copy with someone with out realizing it has some ID numbers at the top.

**Reply by berlef80** [reply]
> When making a plot you can either give it your private keys (this is the default behaviour), or your public keys. Public keys are derived from the private keys but it is public information which you can share with anyone.
> 
> Plotting with public keys can't be done in the GUI right now, only the command line. In theory you can give someone your public keys, they can plot, and then they can send you the file. Farming it requires the private key.

@mariano54 
OK I am coming back to this again, to make sure I understand 100%.  I plot on one machine using GUI and one key.  I then transfer these plots over to a hard drive which is hooked up to a laptop, which uses a different key.  I assume in order for this to work these are two different public keys but have to be the same private key?

If I set up a third key to transfer my won Chia to, should that be a completely new private key unrelated to my other keys?

And finally, if I have a friend who was plotting but is no longer wanting to farm CHIA, is there any way that I can take his plots and farm them on my machine with my keys, or would I have to plot them using his keys?  If the latter then is that possible on my laptop to have two different private keys running at the same time, or do I need a separate machine?

Thank you in advance for your help with this.

**Reply by kazacuuu** [reply]
please help!
If i plotted 100 plots under one key and another 100 plots under a second key but with the same machine full node and i transfered all 200 plots on my harvester machine (not full node ) and then conected my harvester with my full node after generating a third key and start farming .
The quastion is : all my plots are going to be farmed correctly?and if yes ,the rewards for all plots farmed are going to the wallet address i have in my config file?

**Reply by hook66** [reply]
> There are three concerns here:
> 
>     1. Your farming keys should not be in your plotting machines. Instead, plot by using the -f and -p flags (public keys), and then transfer the plots to the machine you will use to farm.
> 
>     2. Your farming keys should not be in your harvester machine, if you have a multi-machine setup. https://github.com/Chia-Network/chia-blockchain/wiki/Farming-on-many-machines
> 
>     3. Your reward address for chia rewards should be a separate key as well, kept in an offline machine. You can generate an address on a different computer, and put this address in the config.yaml (farmer.xch_target_address and pool.xch_target_address), so if your farming machine gets hacked, you don't lose past rewards.
> 
> 
> More technical details here: https://github.com/Chia-Network/chia-blockchain/wiki/Chia-Keys-Architecture

@mariano54 can you just add a wallet address created on a different machine as a reward address (I see two lines in the config file in 1.1.6) with no other change and all the chia rewards will go to that address?

**Comment by af85** [comment]
This is an interesting post. I was wondering if someone could dig a little deeper into point 3 that @mariano54 made. Would it be possible to provide code examples (chia commands) for setting it up, and verifying functionality? I'm wondering if I should spin up a ESXi vm, or use an old raspberry pi to store my wallet/rewards. This is assuming there's no/won't be any support for ledger?

**Reply by hoffmang9** [reply]
Ledger is in process of implement BLS signatures for G2. They're saying "summer" but most expect it a little later. We will promptly be supporting the X and likely supporting the S later.

**Reply by sdbarrish** [reply]
What about QR Codes and Cobo Vault?

**Reply by hoffmang9** [reply]
QR codes will be coming back soon. On the Cobo Wallet we are not sure if/when it will support BLS Signatures on the G2 curve.

**Reply by sdbarrish** [reply]
Thank you kind Sir!

**Reply by sdbarrish** [reply]
> QR codes will be coming back soon. On the Cobo Wallet we are not sure if/when it will support BLS Signatures on the G2 curve.

I sent a request to Cobo as a user of their hard wallet and submitted a feature request.

**Reply by Handri-Kosada** [reply]
Any browser / software wallet that is CHIA compatible like metamask for example?

**Reply by PAVoutsinas** [reply]
Is there any sensitive information / ID numbers in the plot log? I shared a copy with someone with out realizing it has some ID numbers at the top.

**Reply by BitcoinLixin** [reply]
> QR codes will be coming back soon. On the Cobo Wallet we are not sure if/when it will support BLS Signatures on the G2 curve.

Hey Mr Hoffman. Cobo Vault was [rebranded as Keystone](https://twitter.com/BitcoinLixin/status/1399701832167923713). We are open to the discussion of Chia support.

**Comment by 4thFourthHorseman** [comment]
Hello guys,

I have also a question about private keys and I opend a new discussion in the Q&A section. 

I would be thankful, if you could answer my questions.

Thank you very much in advance.

4thHorseman

**Reply by SaintFrantic** [reply]
I've got a question.
So if I add new key and start plotting with it will that key only have the information about this key or will it have all of the information like wallet address and so on?

**Comment by thomann061** [comment]
https://github.com/Chia-Network/chia-blockchain/wiki/Good-Security-Practices-on-Many-Machines#keep-your-wallet-separate

Everytime I update chia-blockchain on ubuntu, my pool and farmer reward addresses get replaced by the wallet keys that are used to farm (I don't want to use those, I want to use a separate wallet reward key). Are there plans to give the option of which reward addresses to use, so I don't have to manually keep replacing them on each update?
